APE_Black_Formatting.rst

```diff
@@ -0,0 +1,286 @@
+base. Currently the style of each subpackage reflects the preferences of the
+maintainers. For instance, the ``time``, ``table``, and ``io.ascii`` subpackages
+allow line lengths of up to 100, while most others use 80 or 88. There are other
+more subtle differences and astropy contributors find this confusing. This has
+come up in the past.
+
+While there are multiple code formatters for Python, the Black Python code
+formatter produces good enough results that it is getting significant traction
+in many open source projects, including: scikit-learn, pytest, tox, Pyramid,
+Django, Hypothesis, SQLAlchemy, virtualenv, pandas, Pillow, and Twisted. While
+adoption by this list of high-profile community-based open source projects does
+not automatically imply that it is a good fit for ``astropy``, it demonstrates
+unequivocally several points:
+
+Technical issues related to the initial transition and subsequent maintenance of
+Black formatting are surmountable. There is plenty of prior art and good
+documentation. Sociological issues related to developer preferences are
+surmountable. Evidence from other projects is that developers either embrace or
+at least accept autoformatting. Developer documentation can provide clear
+instructions to contributors on how to implement Black formatting in their
+development process. Issues with this process are no more difficult (and
+arguably easier) to resolve than enforcing PEP-8 standards.
+
+Therefore, this APE proposes to adopt Black for ``astropy``.
+
+Concerns and mitigations
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Foolish consistency
+~~~~~~~~~~~~~~~~~~~
+A concern that has been highlighted by some members of the community (including,
+initially, one of the co-authors of this APE, and highlighed in detail in `this
+blog post
+<https://luminousmen.com/post/my-unpopular-opinion-about-black-code-formatter>`_)
+is that adopting Black violates the PEP8 principle that "A Foolish Consistency
+is the Hobgoblin of Little Minds". I.e., that by using a code formatter, the
+developer thinks less about readability because they think they can let the
+formatter do this for them. While this concern has some validity, this APE
+explicitly recognizes that **developers and code reviewers are ultimately
+responsible for maintaining a readable code base**. That is, there are many
+choices concerning logic and readability of code that Black does *not* affect,
+and those choices are more important for developers to focus on than mechanistic
+decisions that Black makes about whitespace and newlines.
+
+Undesirable formatting
+~~~~~~~~~~~~~~~~~~~~~~
+On occasion Black's formatting produces undesirable results. The code may look
+very different from expectation due to mathematical or other domain-specific
+conventions, or some specific formatting that conveys meaning may be lost. A
+simple example of this is a constant identity matrix which would typically be
+written as::
+
+    matrix = [[1, 0, 0],
```

Comment:
An even more undesirable formatting (though not with this small, int array, but a large float one would be the examples present in the scipy PR: https://github.com/scipy/scipy/pull/14330)


```
self.a = asarray([1.5750000E-04, 1.6990000E-04, 2.3500000E-04,
                          3.1020000E-04, 4.9170000E-04, 8.7100000E-04,
                          1.7418000E-03, 4.6400000E-03, 6.5895000E-03,
                          9.7302000E-03, 1.4900200E-02, 2.3731000E-02,
                          4.0168300E-02, 7.1255900E-02, 1.2644580E-01,
                          2.0734130E-01, 2.9023660E-01, 3.4456230E-01,
                          3.6980490E-01, 3.6685340E-01, 3.1067270E-01,
                          2.0781540E-01, 1.1643540E-01, 6.1676400E-02,
                          3.3720000E-02, 1.9402300E-02, 1.1783100E-02,
                          7.4357000E-03, 2.2732000E-03, 8.8000000E-04,
                          4.5790000E-04, 2.3450000E-04, 1.5860000E-04,
                          1.1430000E-04, 7.1000000E-05])
```

vs
``` 
        self.a = asarray(
            [
                1.5750000e-04,
                1.6990000e-04,
                2.3500000e-04,
                3.1020000e-04,
                4.9170000e-04,
                8.7100000e-04,
                1.7418000e-03,
                4.6400000e-03,
                6.5895000e-03,
                9.7302000e-03,
                1.4900200e-02,
                2.3731000e-02,
                4.0168300e-02,
                7.1255900e-02,
                1.2644580e-01,
                2.0734130e-01,
                2.9023660e-01,
                3.4456230e-01,
                3.6980490e-01,
                3.6685340e-01,
                3.1067270e-01,
                2.0781540e-01,
                1.1643540e-01,
                6.1676400e-02,
                3.3720000e-02,
                1.9402300e-02,
                1.1783100e-02,
                7.4357000e-03,
                2.2732000e-03,
                8.8000000e-04,
                4.5790000e-04,
                2.3450000e-04,
                1.5860000e-04,
                1.1430000e-04,
                7.1000000e-05,
            ]
        )
```

Reply:
Since this example is not even in astropy, and we would just skip black formatting anyway, I'm not sure what this is telling us.

Reply:
>  and we would just skip black formatting anyway

Then what this APE is about, did I miss the suggestion of not accepting the APE? IMO this is a very much representative example of what's wrong with black, more so than the example in the APE. And there are arrays in astropy, much more than in some other packages that are cited as examples what opted in using it.

Reply:
Is this an example from the test suite?

Reply:
I'm certainly guilty of this, but we should probably minimize hardcoding in arrays for most tests, and instead use more elegant tools like https://github.com/astropy/pytest-arraydiff

Reply:
This is an example from the scipy PR. IMO you cannot ignore mentioning in the APE that neither scipy nor numpy opted in black, exactly because of math and array issues like e.g. this one. 
The scipy PR also shows the bloating up the codebase with whitespaces. 

And they are both very large and diverse contributor base, and much more similar to the one astropy is having, consisting of both scientists and engineers, as opposed to e.g. django.

Reply:
I agree that the numpy and scipy PRs and associated discussion should be mentioned. That does not automatically mean that we need to follow their lead, but mentioning only Django looks like cherry-picking.

Reply:
@bsipocz @hamogu - I added 884b0a1. Is that sufficient?

Reply:
Sounds good to me. Thanks!